src/YOLO/index.js

```diff
@@ -43,108 +43,102 @@ class YOLOBase extends Video {
-        return [];
-      }
-
-      const width = tf.scalar(imageSize);
-      const height = tf.scalar(imageSize);
-      const imageDims = tf.stack([height, width, height, width]).reshape([1, 4]);
-      const boxesModified = tf.mul(boxes, imageDims);
+    const [boxes, scores, classes] = await filterBoxes(allBoxes, boxConfidence, boxClassProbs, this.filterBoxesThreshold);
 
-      const [preKeepBoxesArr, scoresArr] = await Promise.all([
-        boxesModified.data(), scores.data(),
-      ]);
+    // If all boxes have been filtered out
+    if (boxes == null) {
+      return [];
+    }
 
-      const [keepIndx, boxesArr, keepScores] = nonMaxSuppression(
-        preKeepBoxesArr,
-        scoresArr,
-        this.IOUThreshold,
-      );
+    const width = tf.scalar(imageSize);
+    const height = tf.scalar(imageSize);
+    const imageDims = tf.stack([height, width, height, width]).reshape([1, 4]);
+    const boxesModified = tf.mul(boxes, imageDims);
 
-      const classesIndxArr = await classes.gather(tf.tensor1d(keepIndx, 'int32')).data();
+    const [preKeepBoxesArr, scoresArr] = await Promise.all([
+      boxesModified.data(), scores.data(),
+    ]);
 
-      const results = [];
+    const [keepIndx, boxesArr, keepScores] = nonMaxSuppression(
+      preKeepBoxesArr,
+      scoresArr,
+      this.IOUThreshold,
+    );
 
-      classesIndxArr.forEach((classIndx, i) => {
-        const classProb = keepScores[i];
-        if (classProb < this.classProbThreshold) {
-          return;
-        }
+    const classesIndxArr = await classes.gather(tf.tensor1d(keepIndx, 'int32')).data();
 
-        const className = CLASS_NAMES[classIndx];
-        let [y, x, h, w] = boxesArr[i];
+    const results = [];
 
-        y = Math.max(0, y);
-        x = Math.max(0, x);
-        h = Math.min(imageSize, h) - y;
-        w = Math.min(imageSize, w) - x;
+    classesIndxArr.forEach((classIndx, i) => {
+      const classProb = keepScores[i];
+      if (classProb < this.classProbThreshold) {
+        return;
+      }
 
-        const resultObj = {
-          className,
-          classProb,
-          x: x / imageSize,
-          y: y / imageSize,
-          w: w / imageSize,
-          h: h / imageSize,
-        };
+      const className = CLASS_NAMES[classIndx];
+      let [y, x, h, w] = boxesArr[i];
 
-        results.push(resultObj);
-      });
+      y = Math.max(0, y);
+      x = Math.max(0, x);
+      h = Math.min(imageSize, h) - y;
+      w = Math.min(imageSize, w) - x;
 
-      await tf.nextFrame();
-      this.isPredicting = false;
+      const resultObj = {
+        className,
+        classProb,
+        x: x / imageSize,
+        y: y / imageSize,
+        w: w / imageSize,
+        h: h / imageSize,
+      };
 
-      if (callback) {
-        callback(results);
-      }
+      results.push(resultObj);
+    });
 
-      return results;
-    }
-    console.warn('Model has not finished loading');
-    return false;
+    await tf.nextFrame();
```

Comment:
I think we should move this to the beginning of the function, so to only run over the predicions when the `tf` is ready. This made my browser crash.